CMakeLists.txt

```diff
@@ -308,6 +308,20 @@ zephyr_cc_option(-Wpointer-arith)
 # TODO: Archiver arguments
 # ar_option(D)
 
+# Declare MPU userspace dependencies before the linker scripts to make
+# sure the order of dependencies are met
+if(CONFIG_CPU_HAS_MPU AND CONFIG_USERSPACE)
+  if(CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT AND CONFIG_APP_SHARED_MEM )
+    set(APP_SMEM_DEP app_smem_linker)
+  endif()
+  if(CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT AND CONFIG_APPLICATION_MEMORY)
+    set(ALIGN_SIZING_DEP app_sizing_prebuilt linker_app_sizing_script)
+  endif()
+  if(CONFIG_ARM)
+  set(PRIV_STACK_DEP priv_stacks_prebuilt)
+  endif()
+endif()
+
```

Comment:
Each independent patch needs to be in it's own commit.

Re-ordering things can break things in unpredictable and subtle ways,
but a clean history can help to debug these issues.

Reply:
Will try to split it up in more commits.

Reply:
check it now @SebastianBoe

Reply:
You are missing toup markers, but I guess you know that?

Reply:
LGTM